File: python-sdk/src/pvium_sdk/services/webhooks/webhooks.py
from __future__ import annotations

import base64
import hashlib
import hmac
import json
from typing import Any, Dict, Optional


def _b64url_decode(value: str) -> bytes:
    pad_len = (4 - len(value) % 4) % 4
    return base64.urlsafe_b64decode(value + ("=" * pad_len))


def _parse_b64url_json(value: str) -> Optional[Dict[str, Any]]:
    try:
        return json.loads(_b64url_decode(value).decode("utf-8"))
    except Exception:
        return None


def _b64url_hmac_sha256(signing_input: str, secret: str) -> str:
    digest = hmac.new(secret.encode("utf-8"), signing_input.encode("utf-8"), hashlib.sha256).digest()
    return base64.urlsafe_b64encode(digest).decode("utf-8").rstrip("=")


def _safe_equal(a: str, b: str) -> bool:
    return hmac.compare_digest(a, b)


def verifyPviumWebhookToken(token: str, secret: str, options: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    options = options or {}
    parts = token.split(".")
    if len(parts) != 3:
        raise RuntimeError("Invalid Pvium webhook token")

    encoded_header, encoded_payload, encoded_signature = parts
    header = _parse_b64url_json(encoded_header)
    if not header or header.get("alg") != "HS256":
        raise RuntimeError("Unsupported Pvium webhook token algorithm")

    signing_input = f"{encoded_header}.{encoded_payload}"
    secrets = [secret]
    if options.get("allowHashedSecretFallback", True):
        hashed_secret = hashlib.sha256(secret.encode("utf-8")).hexdigest()
        if hashed_secret != secret:
            secrets.append(hashed_secret)

    signature_valid = any(_safe_equal(encoded_signature, _b64url_hmac_sha256(signing_input, candidate)) for candidate in secrets)
    if not signature_valid:
        raise RuntimeError("Invalid Pvium webhook token signature")

    payload = _parse_b64url_json(encoded_payload)
    if not payload:
        raise RuntimeError("Invalid Pvium webhook token payload")

    now = options.get("now")
    if isinstance(now, (int, float)):
        now_seconds = int(now // 1000)
    else:
        import time

        now_seconds = int(time.time())

    exp = payload.get("exp")
    if isinstance(exp, (int, float)) and now_seconds >= int(exp):
        raise RuntimeError("Expired Pvium webhook token")

    expected_event = options.get("expectedEvent")
    if expected_event and payload.get("event") and payload.get("event") != expected_event:
        raise RuntimeError("Pvium webhook token event mismatch")

    return payload


def resolvePviumWebhookPayload(body: Dict[str, Any], secret: str, options: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    options = options or {}
    event = body.get("event") or body.get("type")
    token = body.get("token")

    if not token:
        return {"event": event, "data": body.get("data") or {}}

    merged_options = dict(options)
    if merged_options.get("expectedEvent") is None:
        merged_options["expectedEvent"] = event

    token_payload = verifyPviumWebhookToken(token, secret, merged_options)
    return {
        "event": token_payload.get("event") or event,
        "data": token_payload.get("data") or {},
        "tokenPayload": token_payload,
    }

File: node-sdk/src/webhooks.ts
import HmacSHA256 from "crypto-js/hmac-sha256";
import SHA256 from "crypto-js/sha256";
import Base64 from "crypto-js/enc-base64";
import Utf8 from "crypto-js/enc-utf8";

export interface PviumWebhookTokenPayload<TData = Record<string, unknown>> {
  event?: string;
  data?: TData;
  iat?: number;
  exp?: number;
  [key: string]: unknown;
}

export interface VerifyPviumWebhookTokenOptions {
  expectedEvent?: string;
  now?: Date | number;
  allowHashedSecretFallback?: boolean;
}

export function verifyPviumWebhookToken<
  TData = Record<string, unknown>,
>(
  token: string,
  secret: string,
  options: VerifyPviumWebhookTokenOptions = {},
): PviumWebhookTokenPayload<TData> {
  const parts = token.split(".");
  if (parts.length !== 3) {
    throw new Error("Invalid Pvium webhook token");
  }

  const [encodedHeader, encodedPayload, encodedSignature] = parts;
  const header = parseBase64UrlJson(encodedHeader);
  if (!header || header.alg !== "HS256") {
    throw new Error("Unsupported Pvium webhook token algorithm");
  }

  const signingInput = `${encodedHeader}.${encodedPayload}`;
  const secrets = [secret];
  if (options.allowHashedSecretFallback !== false) {
    const hashedSecret = SHA256(secret).toString();
    if (hashedSecret !== secret) {
      secrets.push(hashedSecret);
    }
  }

  const signatureValid = secrets.some((candidate) =>
    safeEqual(
      encodedSignature,
      hmacSha256Base64Url(signingInput, candidate),
    ),
  );

  if (!signatureValid) {
    throw new Error("Invalid Pvium webhook token signature");
  }

  const payload = parseBase64UrlJson(encodedPayload) as
    | PviumWebhookTokenPayload<TData>
    | null;
  if (!payload) {
    throw new Error("Invalid Pvium webhook token payload");
  }

  const nowSeconds =
    typeof options.now === "number"
      ? Math.floor(options.now / 1000)
      : Math.floor((options.now ?? new Date()).getTime() / 1000);
  if (typeof payload.exp === "number" && nowSeconds >= payload.exp) {
    throw new Error("Expired Pvium webhook token");
  }

  if (
    options.expectedEvent &&
    payload.event &&
    payload.event !== options.expectedEvent
  ) {
    throw new Error("Pvium webhook token event mismatch");
  }

  return payload;
}

export function resolvePviumWebhookPayload<TData = Record<string, unknown>>(
  body: {
    event?: string;
    type?: string;
    token?: string;
    data?: TData;
  },
  secret: string,
  options: VerifyPviumWebhookTokenOptions = {},
): { event?: string; data: TData; tokenPayload?: PviumWebhookTokenPayload<TData> } {
  if (!body.token) {
    return {
      event: body.event ?? body.type,
      data: body.data ?? ({} as TData),
    };
  }

  const tokenPayload = verifyPviumWebhookToken<TData>(body.token, secret, {
    ...options,
    expectedEvent: options.expectedEvent ?? body.event ?? body.type,
  });

  return {
    event: tokenPayload.event ?? body.event ?? body.type,
    data: tokenPayload.data ?? ({} as TData),
    tokenPayload,
  };
}

function hmacSha256Base64Url(value: string, secret: string): string {
  return HmacSHA256(value, secret)
    .toString(Base64)
    .replace(/=/g, "")
    .replace(/\+/g, "-")
    .replace(/\//g, "_");
}

function parseBase64UrlJson(value: string): Record<string, unknown> | null {
  try {
    const base64 = value.replace(/-/g, "+").replace(/_/g, "/");
    const padded = base64.padEnd(
      base64.length + ((4 - (base64.length % 4)) % 4),
      "=",
    );
    const json = Base64.parse(padded).toString(Utf8);
    return JSON.parse(json) as Record<string, unknown>;
  } catch {
    return null;
  }
}

function safeEqual(a: string, b: string): boolean {
  if (a.length !== b.length) return false;

  let diff = 0;
  for (let index = 0; index < a.length; index += 1) {
    diff |= a.charCodeAt(index) ^ b.charCodeAt(index);
  }

  return diff === 0;
}
